chain/cosmos/cosmos_chain.go

```diff
@@ -1197,6 +1193,47 @@ func (c *CosmosChain) StartProvider(testName string, ctx context.Context, additi
 	return nil
 }
 
+func (c *CosmosChain) transformCCVState(ctx context.Context, ccvState []byte, consumerVersion, providerVersion string) ([]byte, error) {
```

Comment:
This func confuses me a bit. Check my logic

- If provider & consumer are the same OR < v3.3.0, no modify
- if provider >3.3.0 and provider is newer than ther consumer, see if newer thjan v4.0.0 and use that. Else use v3.X
- else use the consumer version
- create the consumer genesis json from state. Then use the ics-cd to transfer the genesis to the provider version? This requires `interchain-security-cd` to have an official release rather than using something like gaia. Which means if we wanted to use the ICS v5.0.0-alpha.1 tag, we can't currently 

hmm

Reply:
Almost right :)

If provider >3.3.0 and provider is newer than ther consumer, see if newer thjan v4.0.0 and use that. Else use **v4.0.0 and tell it to transform --to v3.x.x**. This is sadly needed because if I've got a v3.3.0 provider and a v3.1.0 consumer, the v3.3.0 interchain-security-cd binary doesn't know how to transform the genesis to v3.1.0, but the v4.0.0 binary does...

> create the consumer genesis json from state. Then use the ics-cd to transfer the genesis to the provider version?

Almost, use it to transfer the genesis to the consumer version. How this is done depends on the provider version --if the provider is newer, then we need an `ics-cd` that's at least as new as the provider, since it needs to be able to read the new provider's genesis

> This requires interchain-security-cd to have an official release rather than using something like gaia. Which means if we wanted to use the ICS v5.0.0-alpha.1 tag, we can't currently

Correct. I find this extremely annoying. My strong preference would be that consumer chains compile themselves with the `genesis transform` command, which they should be able to get from the ICS version they're linked against. AFAIK no current consumer chain does that, so we do need an `ics-cd` binary. I suppose we could make which docker image it grabs it from configurable, and then if you need, say, an alpha tag, you can build it yourself and configure that image.